stagecoach/price-pusher/price_pusher/core/listener.py

```diff
@@ -0,0 +1,244 @@
+    DurationInSeconds,
+    LatestOrchestratorPairPrices,
+    HumanReadableId,
+)
+from price_pusher.utils.assets import asset_to_pair_id
+from price_pusher.utils.readable_id import generate_human_readable_id
+from pragma.publisher.client import PragmaAPIError
+
+
+logger = logging.getLogger(__name__)
+
+
+class IPriceListener(ABC):
+    """
+    Sends a signal to the Orchestrator when we need to update prices.
+    """
+
+    id: HumanReadableId
+
+    request_handler: IRequestHandler
+    price_config: PriceConfig
+
+    orchestrator_prices: Optional[LatestOrchestratorPairPrices]
+
+    notification_event: asyncio.Event
+    polling_frequency_in_s: DurationInSeconds
+
+    @abstractmethod
+    async def run_forever(self) -> None: ...
+
+    @abstractmethod
+    async def _fetch_all_oracle_prices(self) -> None: ...
+
+    @abstractmethod
+    def set_orchestrator_prices(
+        self, orchestrator_prices: LatestOrchestratorPairPrices
+    ) -> None: ...
+
+    @abstractmethod
+    def _get_most_recent_orchestrator_entry(
+        self, pair_id: str, asset_type: AssetType
+    ) -> Optional[Entry]: ...
+
+    @abstractmethod
+    def _notify(self) -> None: ...
+
+    @abstractmethod
+    def _log_listener_spawning(self) -> None: ...
+
+    @abstractmethod
+    async def _does_oracle_needs_update(self) -> bool: ...
+
+    @abstractmethod
+    def _new_price_is_deviating(self, pair_id: str, new_price: int, oracle_price: int) -> bool: ...
+
+    @abstractmethod
+    def _oracle_entry_is_outdated(
+        self, pair_id: str, oracle_entry: Entry, newest_entry: Entry
+    ) -> bool: ...
+
+
+class PriceListener(IPriceListener):
+    def __init__(
+        self,
+        request_handler: IRequestHandler,
+        price_config: PriceConfig,
+        polling_frequency_in_s: DurationInSeconds,
+    ) -> None:
+        self.id = generate_human_readable_id()
+        self.request_handler = request_handler
+        self.price_config = price_config
+
+        self.oracle_prices = {}
+        self.orchestrator_prices = None
+
+        self.notification_event = asyncio.Event()
+        self.polling_frequency_in_s = polling_frequency_in_s
+
+        self._log_listener_spawning()
+
+    async def run_forever(self) -> None:
+        """
+        Main loop responsible of:
+            - fetching the latest oracle prices
+            - checking if the oracle needs update
+            - pushing notification to the orchestration if it does.
+        """
+        last_fetch_time = -1
+        while True:
+            current_time = asyncio.get_event_loop().time()
+            if current_time - last_fetch_time >= self.polling_frequency_in_s:
+                await self._fetch_all_oracle_prices()
+                last_fetch_time = current_time
+            if await self._does_oracle_needs_update():
+                self._notify()
+                last_fetch_time = -1
+            # Check every second if the oracle needs an update
+            await asyncio.sleep(1)
```

Comment:
I don't think we need that sleep since the previous async operations will probably take longer than 1s anyway

Reply:
removing the sleep mess things up here - it's taking a loooot of the available resources I think.
for example, it makes some requests (1/2) from the Poller fail for some reasons and some other strange things

I didn't removed it but lowered the sleep to 0.1